core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java

```diff
@@ -87,4 +100,44 @@ public static TypeAnalyzer createTestingTypeAnalyzer(PlannerContext plannerConte
                         new TablePropertyManager(),
                         new AnalyzePropertyManager()));
     }
+
+    private static class QueryScopedCachedTypeAnalyzer
+    {
+        private final Cache<List<NodeRef<Expression>>, Map<NodeRef<Expression>, Type>> typesCache = buildNonEvictableCache(
+                CacheBuilder.newBuilder()
+                        .weakValues()
```

Comment:
`List<NodeRef<Expression>>` is quite unusual cache key.
If i ask for a type for a list of 100 elements, and then i re-ask with only 99  of them, i get a miss.

also, list + weak values is an unnatural choice. Lists are used so often, that it's hard to track their lifecycle.